### PR TITLE
AR multiparameter

### PR DESCRIPTION
**Pull request by tamird**
closes https://github.com/attr-encrypted/attr_encrypted/pull/80.

@sbfaulkner @danielnc @billymonk @saghaulor

[Review on Reviewable](https://reviewable.io/reviews/attr-encrypted/attr_encrypted/115)

### CODE REVIEW

**Comment by saghaulor** [comment]
@tamird wow nice work!

**Comment by saghaulor** [comment]
@tamird as I've merged your cleanup work this PR can't auto merge. Please rebase.

**Comment by tamird** [comment]
@saghaulor rebased!

**Comment by tamird** [comment]
should be green now

**Comment by tamird** [comment]
@saghaulor this good to merge?

**Comment by saghaulor** [comment]
@tamird I'm still reviewing this code. Trying to wrap my head around it. But please know that I haven't forgotten about it. Thank you for the reminder.

**Comment by saghaulor** [comment]
@tamird I think I understand what is going on. I've tested this code out on my app andI've hit a stack trace. I'm trying to narrow down if it's something specific to our implementation or if it's an issue with this code.

I'll keep digging and keep you posted.

**Comment by tamird** [comment]
okay, let me know.

**Comment by tamird** [comment]
@saghaulor ping

**Comment by tamird** [comment]
@saghaulor this is green ...

**Comment by saghaulor** [comment]
@tamird I believe that I understand the intent. We're doing this because otherwise when the data is saved to the db, the data is mutated into a string? 

I have some concerns. The test is confusing. Is it possible to simplify? I wonder if we should be defining type_for_attribute_with_attr_encrypted when type_for_attribute may not be defined? Lastly, I'm concerned about removing the before_typecast part, I removed it previously in an experiment and it seemed to have broke the way the accessors decrypt strings that are rendered in views, possibly, in other places as well. I'll try to run this commit against my app and see if everything is working as expected.

Thanks again for all the hard work that you've been putting in.

**Comment by tamird** [comment]
@saghaulor tried to make the test clearer. what do you think?

**Comment by tamird** [comment]
> I have some concerns. The test is confusing. Is it possible to simplify?

Attempted to address this.
> I wonder if we should be defining type_for_attribute_with_attr_encrypted when type_for_attribute may not be defined?

Yeah, that's not ideal, but I believe it's harmless.
> Lastly, I'm concerned about removing the before_typecast part, I removed it previously in an experiment and it seemed to have broke the way the accessors decrypt strings that are rendered in views, possibly, in other places as well. I'll try to run this commit against my app and see if everything is working as expected.

This makes me sad...I don't think we should hold up merging this functionality over this. If an issue comes up, we'll fix it. Until we have tests covering the expected behaviour this is nothing more than FUD.

**Comment by saghaulor** [comment]
@tamird I respectfully disagree. Given the knowledge that this gem is used in literally thousands of applications, a certain amount of uncertainty and doubt are healthy and responsible. As such we should be mindful of how changes to it might impact others. I will conduct what I believe to be due diligence and if everything checks out I'll merge it. 

Moreover, it is my understanding that there has been a work around for this functionality. As such I have been trying to understand if the current solution is the best solution, and/or if it's even appropriate.

To your point about tests, I agree that tests should be the basis of how we determine the behavior of the code. In this particular scenario, when I removed the aforementioned code, the impact was not obvious. All tests passed, it was only when the code interacted with the view in Rails that the impact was noticeable. How do you suggest that we approach that? Do we include an entire Rails application with UI level tests as part of our test suite?

**Comment by tamird** [comment]
> To your point about tests, I agree that tests should be the basis of how we determine the behavior of the code. In this particular scenario, when I removed the aforementioned code, the impact was not obvious. All tests passed, it was only when the code interacted with the view in Rails that the impact was noticeable. How do you suggest that we approach that? Do we include an entire Rails application with UI level tests as part of our test suite?

Were you able to reproduce the issue? I think we probably won't need to embed a full rails app - I'm sure the minimum test case is reducible (this gem doesn't muck with ActionView, and the tests already bring in ActiveRecord).

**Comment by tamird** [comment]
Ping.

**Comment by tamird** [comment]
Ping!

**Comment by tamird** [comment]
@rcook are you the new maintainer? Could you take a look, please?

**Comment by rcook** [comment]
@tamird: I am no longer a maintainer, I'm afraid.

**Comment by tamird** [comment]
Whose leg do must I hump to get someone to review this?

**Comment by tamird** [comment]
rebased this again. for the love of god, PTAL?

**Comment by rcook** [comment]
@tamird : As I mentioned above I am no longer a maintainer. Perhaps you should e-mail one or more of @sbfaulkner, @billymonk or @saghaulor directly?

-Richard

**Comment by tamird** [comment]
Rebased...again.

**Comment by rcook** [comment]
@tamird : Have you tried contacting @sbfaulkner , @billymonk or @saghaulor directly?

**Comment by sbfaulkner** [comment]
I too have not looked at this in quite a while... @saghaulor have you considered this PR?

**Comment by saghaulor** [comment]
@sbfaulkner I have looked at it. I have some concerns. I plan to express them soon. I wanted to look deeper into this PR before I started making comments.

@tamird thank you for your effort and your patience.

File: lib/attr_encrypted/adapters/active_record.rb [hunk 1]
```diff
@@ -43,32 +50,47 @@ def attributes_with_attr_encrypted=(*args)
...
+                self.class.decrypt(attribute, public_send("#{attribute_options[:attribute]}_was"), evaluated_options)
+              end
             end
+          end
+
+          def reflect_on_aggregation_with_attr_encrypted(attribute)
+            encrypted_reflections[attribute.to_sym] || reflect_on_aggregation_without_attr_encrypted(attribute)
+          end
 
-            alias_method "#{attr}_before_type_cast", attr
```
Comment (saghaulor) [inline lib/attr_encrypted/adapters/active_record.rb]: Is it necessary to remove this? I think it makes sense to do the decryption before typecasting.

File: lib/attr_encrypted/adapters/active_record.rb [hunk 2]
```diff
@@ -43,32 +50,47 @@ def attributes_with_attr_encrypted=(*args)
           end
         end
 
+        # Saves the attr class wrappers, so AR can infer the correct type later
+        def encrypted_reflections
```
Comment (saghaulor) [inline lib/attr_encrypted/adapters/active_record.rb]: Are the reflections encrypted? I don't believe they are. If so, this name is misleading.

File: lib/attr_encrypted/adapters/active_record.rb [hunk 3]
```diff
@@ -2,6 +2,9 @@
   module AttrEncrypted
     module Adapters
       module ActiveRecord
+        # Allows AR to infer the proper type for the column
+        EncryptedReflection = Struct.new(:klass)
```
Comment (saghaulor) [inline lib/attr_encrypted/adapters/active_record.rb]: Are the reflections encrypted? I don't believe they are. If so, this name is misleading.

**Comment by saghaulor** [comment]
@tamird I've added some comments regarding concerns that I have with this PR. Thank you for efforts.

File: lib/attr_encrypted/adapters/active_record.rb (hunk 3 above)
Comment (tamird) [inline lib/attr_encrypted/adapters/active_record.rb]: Updated the name to remove ambiguity.

File: lib/attr_encrypted/adapters/active_record.rb (hunk 2 above)
Comment (tamird) [inline lib/attr_encrypted/adapters/active_record.rb]: Updated the name to remove ambiguity.

File: lib/attr_encrypted/adapters/active_record.rb (hunk 1 above)
Comment (tamird) [inline lib/attr_encrypted/adapters/active_record.rb]: I don't quite follow - can you show me a test case that would highlight the change in behaviour?

**Pull request closed without merging**